src/charm_state.py

```diff
@@ -65,16 +127,68 @@ def __init__(self, msg: str):
         self.msg = msg
 
 
+def _valid_storage_size_str(size: str) -> bool:
+    """Validate the storage size string.
+
+    Args:
+        size: Storage size string.
+
+    Return:
+        Whether the string is valid.
+    """
+    # Checks whether the string confirms to using the KiB, MiB, GiB, TiB, PiB,
+    # EiB suffix for storage size as specified in config.yaml.
+    valid_suffixes = ["KiB", "MiB", "GiB", "TiB", "PiB", "EiB"]
+    valid = any(suffix == size[-3:] for suffix in valid_suffixes)
+
+    valid = valid or size[:-3].isdigit()
+    return valid
+
+
 class CharmConfig(BaseModel):
-    """Charm configuration.
+    """General charm configuration.
+
+    Some charm configurations are grouped into other configuration models.
 
     Attributes:
-        runner_storage: Storage to be used as disk for the runner.
+        path: GitHub repository path in the format '<owner>/<repo>', or the GitHub organization
+            name.
+        token: GitHub personal access token for GitHub API.
+        reconcile_interval: Time between each reconciliation of runners.
+        denylist: List of IPv4 to block the runners from accessing.
+        dockerhub_mirror: Private docker registry as dockerhub mirror for the runners to use.
+        openstack_clouds_yaml: The openstack clouds.yaml configuration.
     """
 
-    runner_storage: RunnerStorage
+    path: GithubPath
+    token: str
+    reconcile_interval: int
+    denylist: list[FirewallEntry]
+    dockerhub_mirror: str | None
     openstack_clouds_yaml: dict | None
 
+    @classmethod
+    def _parse_denylist(cls, charm: CharmBase) -> list[str]:
+        denylist_str = charm.config.get("denylist", "")
+
+        entry_list = [entry.strip() for entry in denylist_str.split(",")]
+        denylist = [FirewallEntry.decode(entry) for entry in entry_list if entry]
+        return denylist
+
+    @classmethod
+    def _parse_dockerhub_mirror(cls, charm: CharmBase) -> str | None:
+        dockerhub_mirror = charm.config.get("dockerhub-mirror") or None
+
+        dockerhub_mirror_url = urlsplit(dockerhub_mirror)
```

Comment:
I have tested that case, it would just return a empty object.
```
>>> urlsplit(None)
SplitResultBytes(scheme=b'', netloc=b'', path=b'', query=b'', fragment=b'')
```